feat(env): add convenience apis for common scripting needs

andrewbrey opened the pull request:
Borrowing from my own project which I _started_ prior to learning about yours (https://github.com/andrewbrey/desh), I have found that there are a few things I end up repeating in every almost single script, so it's really convenient to have them just be baked into the runtime api. This PR adds some of these:

- Globally available variables for the `HOME` and `USER` environment vars since they are so commonly accessed
- Globally available variables which are the equivalent of the numerically indexed script arguments, i.e. `$1`, `$2`, etc. We already have `$.args` but it's very convenient to be able to access common values in that array without hassle
- Most importantly, convenience wrappers around `$` for extracting only the exit code, the trimmed stdout, or the trimmed stderr. In this PR, these are provided as:
	- the `$s` template tag which extracts only the exit status of a command
	- the `$o` template tag which extracts only the trimmed child process stdout
	- the `$e` template tag which extracts only the trimmed child process stderr

Let me know if you have questions or want changes! Hopefully you like these kinds of convenience additions to the runtime API :)

c4spar commented:
hey @andrewbrey thx for your next contribution!


> * Globally available variables for the HOME and USER environment vars since they are so commonly accessed

I'm currently not sure if i want to add this globals. I think it would be better if we don't add to much to the global scope. Currently there is no `std/os` module and the `std/node/os` module is not done yet but at some point i want to add an `os` module that provides methods like `homedir`, `username` or `userinfo` etc... Meanwhile we could create our own `os` module with only a `homedir` and `username` method. What do you think?

> * Globally available variables which are the equivalent of the numerically indexed script arguments, i.e. $1, $2, etc. We already have $.args but it's very convenient to be able to access common values in that array without hassle

I looked at `zx` and they parse the arguments with minimist initially and make them globally available as `argv` variable. we could do something similar and parse the arguments with the `std/flags` module instead of adding indexed variables.

> * Most importantly, convenience wrappers around $ for extracting only the exit code, the trimmed stdout, or the trimmed stderr. In this PR, these are provided as:
>    * the $s template tag which extracts only the exit status of a command
>    * the $o template tag which extracts only the trimmed child process stdout
>    * the $e template tag which extracts only the trimmed child process stderr

Currently i'm working on a `Process` class that implements `PromiseLike`, `AsyncIterable`, `Deno.Reader`, `Deno.Writer`. We could also implement `Deno.Process` but instead of returning `Uint8Array`'s we could return `string`s. The class would include the functionality of `$e`, `$o` and `$s` and could look something like this:

```ts
declare class Process
  implements
    PromiseLike<ProcessOutput>,
    AsyncIterable<string>,
    Deno.Reader,
    Deno.Writer {
  readonly rid: number;
  readonly pid: number;
  readonly stdin: (Deno.Writer & Deno.Closer) | null;
  readonly stdout: (Deno.Reader & Deno.Closer) | null;
  readonly stderr: (Deno.Reader & Deno.Closer) | null;
  [Symbol.asyncIterator](): AsyncIterator<string>;
  then<T, S>(
    resolve: (value: ProcessOutput) => T | Promise<T>,
    reject?: (error: Error) => S | Promise<S>,
  ): Promise<T | S>;
  read(p: Uint8Array): Promise<number | null>;
  write(p: Uint8Array): Promise<number>;
  status(): Promise<Deno.ProcessStatus>;
  statusCode(): Promise<number>; // or exitCode
  output(): Promise<string>;
  stderrOutput(): Promise<string>;
  copy(process: Process): Process;
  close(): void; // or pipe or pipeThrough
  kill(signo: number): void;
  noThrow(): Process;
}
```

```ts
// get stdout & stderr output:
const output: ProcessOutput = await $`echo foo`;

// get only stdout output:
const output: string = await $`echo foo`.output();

// get only stderr output:
const errorMessage: string = await $`echo foo`.stderrOutput();

// get only exit code:
const statusCode: number = await $`echo foo`.statusCode();

// don't throw errors:
const output: ProcessOutput = await $`echo foo`.noThrow();
// can be chained:
const { code, signal } = await $`echo foo`.noThrow().status();

// iterate over each line:
for (const file of $`ls -1`) {
  console.log("file:", file);
}

// pipe process output to a second process:
const count: number = parseInt($`ls -1`.copy($`wc -l`));
// same as:
const count: number = parseInt($`ls -1 | wc -l`);

// pipe output to a file:
const file = await Deno.open("file.txt");
await $`echo foo`.copy(file);
file.close();
// same as:
const file = await Deno.open("file.txt");
await Deno.copy($`echo foo`, file);
file.close();
```

What do you think about this api?

And it would be cool if you can create separate PR's for each feature 🙂. Because reviewing is much easier if each feature/bugfix has it's own branch, because than we can better discuss and add changes to the PR and i can squash the branch to keep the history clean and it's easier to create a changelog from a clean git history. Will add an contribution guideline soon 😉

andrewbrey commented:
Thanks for the review! First off, re:

> And it would be cool if you can create separate PR's for each feature slightly_smiling_face. Because reviewing is much easier if each feature/bugfix has it's own branch, because than we can better discuss and add changes to the PR and i can squash the branch to keep the history clean and it's easier to create a changelog from a clean git history. Will add an contribution guideline soon wink

Sorry if this felt like too many separate thoughts in a single PR - in my head, these were each "small ergonomics-focused additions to the existing runtime" so a single PR felt right to me. Happy to break up contributions however is preferred in the future and if/when you have a contributing doc, I'll be happy to comply with whatever is asked there :)

Next, re: `HOME`/`USER`/`$1 ...$9`, I understand the preference to keep the globals to a minimum - it's a bit of an indistinct line though because one might ask why `$` is a global anyway? And why should `$` have colors attached to it, but not (for instance) `cd`. I know that the original inspiration was `zx` which uses a global for these things, but honestly, I think it's not all that helpful in the end since you either need to use the named imports from the runtime module _or_ you need to include a triple-slash type reference comment (if you care about the authoring experience at all at least)..either way, you end up typing a line at the top of your module, at which point it begs the question...why bother with _any_ globals? Just a thought I've had as I've done a lot of automation with `zx`...

With respect to this PR, I think it would just make sense for me to remove all globals I introduced (I can add them back for myself via an import of my own and the only thing I lose is the global availability...which as mentioned, I'm not too gutted by :smile: )

Now, re: the new api for process - I think for the most part this will be great. I do think that the template tags I introduced as wrappers of `exec` are a slightly better authoring experience for their usecase though. For instance, in your new API, to get the trimmed stdout I would write:

```ts
const result = await $`pwd`.output().trim()
```

which I _could_ already do today with:

```ts
const result = (await $`pwd`).stdout.trim()
```

which is actually fewer characters. The same with the wrapper template tag `$o` is:

```ts
const result = await $o`pwd`
```

So, basically, I sort of feel like, if you are opposed to the added template tags to tackle this piece of the ergonomics, I'm not sure that the new API really improves on the state of things as they are now (specifically with respect to the common need to just get the stdout, stderr, and exit code). Other new parts of this `Process` wrapper are going to be great, like the chainability the ability to pipe to a file.

What if you make the new API slightly less verbose with something like:

```ts

declare class Process
  //...clipped for brevity
  set nothrow(): Process
  // these could automatically apply the internal state 
  // change which the nothrow setter would apply, and they
  // trim output
  get exitCode(): Promise<number>;
  get stdout(): Promise<string>;
  get stderr(): Promise<string>;
}
```

so that you could go:

```ts
const code = await $`exit 2;`.exitCode // --> 2
const out = await $`echo hi;`.stdout // --> hi
const err = await $`echo hi >&2`.stderr // --> hi
```

This achieves a lot of the brevity I wanted out of the template tags but keeps it all encapsulated inside of the `Process` class...thoughts?

andrewbrey commented:
I've remove `HOME`, `USER`, and `$1...$9` :heavy_check_mark:

c4spar commented:
Hi @andrewbrey, sorry for the delay, I've been busy.

I agree it is inconsistent to have the colors on `$` and i am open to move them to a `colors` module.
Can you remove the `homedir` and `username` functions too? if you want you can open a second pr to introduce an `os` module with an `os.homedir()` and `os.username()` function. i think it's more consistent with the other modules to have it on a namespace. 
But do we really need the `envExists` and `envMissing`? why not just `Deno.env.get` and `!Deno.env.get`? I think i would also remove this functions and i think normally it shouldn't be necessary to trim an env variable.
With the rest i'm fine to merge. i will think more about the new api what makes most sense and i will open a new issue for this.

andrewbrey commented:
@c4spar Sure thing, happy to defer to your preferences on all of this. If you think more about it and decide that you would prefer to iterate more privately on the API of the `Process` interface, and for now, not even merge in these remaining changes to add `exec` wrappers, please don't hesitate to just close this PR. This is your library, feel free to enforce your preferences :)

Also no worries on the delay, I've been busy too :+1:

andrewbrey commented:
@c4spar is there anything additional you wanted from me on this PR? Please let me know if there are outstanding change requests to address which I missed :)
